### PR TITLE
Make fstype configurable in external provisioner

### PR DESCRIPTION
**Pull request by humblec**
At present the fstype is set to `ext4` if nothing is passed in storage-class.
However a SP could prefer to have different fstype for many reasons for their
driver/volumes. This patch enables SP who is using the external-provisioner
to choose the  fstype which they want to have it.

Fix# https://github.com/kubernetes-csi/external-provisioner/issues/328

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

> /kind bug

**What this PR does / why we need it**:

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
 The `defaultFStype` arg is no longer default to "ext4" filesystem type. This arg is available for SP to use when they deploy provisioner sidecar. Admins can also specify this `fstype` via storage class parameter. If `fstype` is set in storage class parameter, it will be used. The sidecar arg is only checked if fstype is not set in the SC param.
```

### CODE REVIEW

**Comment by humblec** [comment]
@jsafrane @msau42 ptal .

File: cmd/csi-provisioner/csi-provisioner.go [hunk 1]
```diff
@@ -89,6 +90,8 @@ func main() {
 	flag.Var(utilflag.NewMapStringBool(&featureGates), "feature-gates", "A set of key=value pairs that describe feature gates for alpha/experimental features. "+
 		"Options are:\n"+strings.Join(utilfeature.DefaultFeatureGate.KnownFeatures(), "\n"))
 
+	flag.StringVar(&provisionercontroller.DefaultFSType, "default-fstype", "ext4", "Specify the filesystem type of the volume. If not specified,  external-provisioner will set default as `ext4`.")
```
Comment (davidz627) [inline cmd/csi-provisioner/csi-provisioner.go]: If we're going to do this I would prefer we follow the style of defining the other flags like `metricsPath`

Comment (humblec) [reply cmd/csi-provisioner/csi-provisioner.go]: @davidz627 the thing to note here is, `defaultFStype` is already defined here https://github.com/kubernetes-csi/external-provisioner/blob/master/pkg/controller/controller.go#L119 unlike other variables in this section like `metricsPath`. So, isnt it good to be here ?

Comment (davidz627) [reply cmd/csi-provisioner/csi-provisioner.go]: that can be changed, it's just a variable

Comment (humblec) [reply cmd/csi-provisioner/csi-provisioner.go]: @davidz627 this comment is addressed. PTAL 
@jsafrane 
@msau42

File: cmd/csi-provisioner/csi-provisioner.go [hunk 2]
```diff
@@ -73,6 +73,8 @@ var (
 	metricsAddress = flag.String("metrics-address", "", "The TCP network address where the prometheus metrics endpoint will listen (example: `:8080`). The default is empty string, which means metrics endpoint is disabled.")
 	metricsPath    = flag.String("metrics-path", "/metrics", "The HTTP path where prometheus metrics will be exposed. Default is `/metrics`.")
 
+	defaultFSType = flag.String("default-fstype", "ext4", "Specify the filesystem type of the volume. If not specified,  external-provisioner will set default as `ext4`.")
```
Comment (xing-yang) [inline cmd/csi-provisioner/csi-provisioner.go]: This would still force the driver to specify a default fstype ahead of time.  Can we just let driver determine and set this value internally?

Comment (Madhu-1) [reply cmd/csi-provisioner/csi-provisioner.go]: https://github.com/kubernetes-csi/external-provisioner/issues/328#issuecomment-528442910

Comment (pohly) [reply cmd/csi-provisioner/csi-provisioner.go]: We could introduce `--default-fstype=none` as a backwards-compatible way of letting a CSI driver specify "don't set a default in external-provisioner, I'll handle that myself". "ext4" would still be the default, so nothing changes for drivers that don't use this new option.

Comment (humblec) [reply cmd/csi-provisioner/csi-provisioner.go]: Yeah,  defaulting to `ext4` would be the way for now. Isnt this PR good to go then? @jsafrane @msau42 @xing-yang ?

Comment (pohly) [reply cmd/csi-provisioner/csi-provisioner.go]: That's not quite what I meant. We should default to `ext4` to avoid the need for a major release, but *also* allow a CSI driver to use the modified external-provisioner without any defaulting. We can enable the latter with `--default-fstype=none`.

File: pkg/controller/controller.go [hunk 3]
```diff
@@ -496,7 +497,7 @@ func (p *csiProvisioner) ProvisionExt(options controller.ProvisionOptions) (*v1.
 		return nil, controller.ProvisioningFinished, fmt.Errorf("fstype specified in parameters with both \"fstype\" and \"%s\" keys", prefixedFsTypeKey)
 	}
 	if len(fsType) == 0 {
-		fsType = defaultFSType
+		fsType = p.defaultFSType
```
Comment (pohly) [inline pkg/controller/controller.go]: Here's the implementation of `--default-fstype=none`:

```
if fsType == "" && p.defaultFSType != "none" {
    fsType = p.defaultFSType
}
```

Comment (humblec) [reply pkg/controller/controller.go]: Done.

File: cmd/csi-provisioner/csi-provisioner.go (hunk 2 above)
Comment (humblec) [reply cmd/csi-provisioner/csi-provisioner.go]: Done..

Comment (xing-yang) [reply cmd/csi-provisioner/csi-provisioner.go]: Can you add a brief explanation of what `none` means in the help message and also add a unit test for `--default-fstype=none`?

Comment (msau42) [reply cmd/csi-provisioner/csi-provisioner.go]: I am actually thinking of releasing a 2.0 this cycle.  There are a number of 2.0 changes that I want to target and I think we have gathered enough that it's time to do them all.

So I think we can actually make the change to remove the defaulting altogether and not have a new argument for it.

(I do want to release a 1.6.0 first, so let's not merge this until I cut a 1.6.0)

Comment (xing-yang) [reply cmd/csi-provisioner/csi-provisioner.go]: Removing the defaulting will be even better.

Comment (humblec) [reply cmd/csi-provisioner/csi-provisioner.go]: > I am actually thinking of releasing a 2.0 this cycle. There are a number of 2.0 changes that I want to target and I think we have gathered enough that it's time to do them all.
> 
> So I think we can actually make the change to remove the defaulting altogether and not have a new argument for it.
> 
> (I do want to release a 1.6.0 first, so let's not merge this until I cut a 1.6.0)

@msau42 iiuc, the suggestion here is to remove the `defaultFSType` arg completely from this PR and keep it ready for `2.0` , Isnt it ?

Comment (msau42) [reply cmd/csi-provisioner/csi-provisioner.go]: sorry to flip back and forth on this. We discussed this in the csi standup today and we may still want to provide an option to set a default so that we can do fstype source/target matching enforcement for cloning and potentially snapshots. cc @jsafrane @xing-yang @j-griffith

Comment (humblec) [reply cmd/csi-provisioner/csi-provisioner.go]: @msau42 done.. nw :).. ptal .

File: pkg/controller/controller_test.go [hunk 4]
```diff
@@ -66,6 +66,7 @@ var (
 
 	driverNameAnnotation = map[string]string{annStorageProvisioner: driverName}
 	translatedKey        = "translated"
+	defaultfsType        = "ext4"
```
Comment (msau42) [inline pkg/controller/controller_test.go]: Add a test case with the "none" option?

File: cmd/csi-provisioner/csi-provisioner.go (hunk 2 above)
Comment (msau42) [reply cmd/csi-provisioner/csi-provisioner.go]: Do we want to wait for a 2.0 so that we can add the option without a default?

Will it be confusing if we add an option now and immediately deprecate it and change it in the next release?

Comment (msau42) [reply cmd/csi-provisioner/csi-provisioner.go:76]: 1.6 is released. And I expect this next release to be a 2.0. So let's go ahead and add the new option without a default. I have mentioned this in the deprecation notes.

Comment (humblec) [reply cmd/csi-provisioner/csi-provisioner.go:76]: @msau42 sure . thanks!

**Comment by msau42** [comment]
@humblec can you rebase? We're ready to merge this.

**Comment by humblec** [comment]
> @humblec can you rebase? We're ready to merge this.

Sure @msau42 .. just went out of radar due to other clutter of PRs..

File: cmd/csi-provisioner/csi-provisioner.go [hunk 5]
```diff
@@ -74,6 +71,8 @@ var (
...
```
Comment (msau42) [inline cmd/csi-provisioner/csi-provisioner.go:74]: Can we make this opt-in so that plugins don't need a "none" value?

**Comment by msau42** [comment]
Can you also add a release note?

File: cmd/csi-provisioner/csi-provisioner.go (hunk 5 above)
Comment (xing-yang) [reply cmd/csi-provisioner/csi-provisioner.go:74]: Can you add a comment here about "none"?  Otherwise driver developer won't know that they can set "none" here unless if they read the code.

Comment (xing-yang) [reply cmd/csi-provisioner/csi-provisioner.go:74]: @msau42 do you mean we should allow plugin to leave this as "" instead of setting to "none" to indicate there is no default fstype?

Comment (msau42) [reply cmd/csi-provisioner/csi-provisioner.go:74]: Correct

**Comment by humblec** [comment]
@msau42 @xing-yang I have to admit that, due to various comments in different times, I am confused a bit here. however the lastpatchset is carved out in such a way, the `defaultFStype` arg is available for SP's to use when they deploy sidecar with a default value `"" `. If not set, the fsType value will be unfilled  Admins can also specify this fstype via SC param. If both SC param and provisioner sidecar arg has been filled , the preference goes to the sidecar arg. 

Is this the desired thing we want to achieve? Did I miss anything here?

**Comment by xing-yang** [comment]
> If both SC param and provisioner sidecar arg has been filled , the preference goes to the sidecar arg.

I'm looking at the code in this PR.  It is not exactly what you described above.  If fstype is set in SC param, it will be used.  The sidecar arg is only checked if fstype is not set in the SC param.  This looks right to me.

**Comment by humblec** [comment]
> > If both SC param and provisioner sidecar arg has been filled , the preference goes to the sidecar arg.
> 
> I'm looking at the code in this PR. It is not exactly what you described above. If fstype is set in SC param, it will be used. The sidecar arg is only checked if fstype is not set in the SC param. This looks right to me.

Ah.. exactly. Corrected my explanation too!

**Comment by xing-yang** [comment]
Can you add this description to the release note?

**Comment by humblec** [comment]
> Can you add this description to the release note?

I have added it. Please let me know if it need to be modified @xing-yang

**Comment by xing-yang** [comment]
Can you move the release notes to the "Does this PR introduce a user-facing change?" section?

**Comment by humblec** [comment]
> Can you move the release notes to the "Does this PR introduce a user-facing change?" section?

@xing-yang may be both can be set, I mean `release-note` and also `user-facing-*` section ?

File: cmd/csi-provisioner/csi-provisioner.go [hunk 6]
```diff
@@ -74,6 +71,8 @@ var (
 	metricsAddress = flag.String("metrics-address", "", "The TCP network address where the prometheus metrics endpoint will listen (example: `:8080`). The default is empty string, which means metrics endpoint is disabled.")
 	metricsPath    = flag.String("metrics-path", "/metrics", "The HTTP path where prometheus metrics will be exposed. Default is `/metrics`.")
 
+	defaultFSType = flag.String("default-fstype", "", "Specify the filesystem type of the volume. If not specified,  external-provisioner will not set any default filesystem type.")
```
Comment (msau42) [inline cmd/csi-provisioner/csi-provisioner.go:74]: How about something like:
```
The default filesystem type of the volume to provision when fstype is unspecified in the StorageClass. If the default is not set and fstype is unset in the StorageClass, then no fstype will be set.
```

File: pkg/controller/controller.go [hunk 7]
```diff
@@ -518,8 +519,8 @@ func (p *csiProvisioner) ProvisionExt(options controller.ProvisionOptions) (*v1.
 	if fsTypesFound > 1 {
 		return nil, controller.ProvisioningFinished, fmt.Errorf("fstype specified in parameters with both \"fstype\" and \"%s\" keys", prefixedFsTypeKey)
 	}
-	if len(fsType) == 0 {
-		fsType = defaultFSType
+	if fsType == "" && p.defaultFSType != "" {
```
Comment (msau42) [inline pkg/controller/controller.go:522]: Can you add unit tests that test the various combinations?
- StorageClass set + default set
- StorageClass unset + default set
- StorageClass set + default unset
- StorageClass unset + default unset

**Comment by humblec** [comment]
Running tool: /usr/local/go/bin/go test -timeout 30s github.com/kubernetes-csi/external-provisioner/pkg/controller -run ^(TestProvision)$

ok  	github.com/kubernetes-csi/external-provisioner/pkg/controller	0.040s

/usr/local/go/bin/go test -timeout 30s github.com/kubernetes-csi/external-provisioner/pkg/controller -run ^(TestFSTypeProvision)$

ok  	github.com/kubernetes-csi/external-provisioner/pkg/controller	(cached)

File: cmd/csi-provisioner/csi-provisioner.go (hunk 6 above)
Comment (humblec) [reply cmd/csi-provisioner/csi-provisioner.go]: Sure. Done.

**Comment by humblec** [comment]
@msau42 @jsafrane @xing-yang any thoughts on why few of the existing tests ( For ex: TestProvision) are failing randomly when package-level test is triggered ?

File: pkg/controller/controller_test.go [hunk 8]
```diff
@@ -1709,9 +1838,12 @@ func runProvisionTest(t *testing.T, k string, tc provisioningTestcase, requested
 	clientSet := fakeclientset.NewSimpleClientset(tc.clientSetObjects...)
 
 	pluginCaps, controllerCaps := provisionCapabilities()
-	csiProvisioner := NewCSIProvisioner(clientSet, 5*time.Second, "test-provisioner", "test", 5, csiConn.conn,
-		nil, provisionDriverName, pluginCaps, controllerCaps, supportsMigrationFromInTreePluginName, false, csitrans.New(), nil, nil, nil, nil, nil, tc.withExtraMetadata)
 
+	if tc.skipDefaultFSType {
+		defaultfsType = ""
```
Comment (msau42) [inline pkg/controller/controller_test.go:1843]: I think the unit test is failing because this is modifying a global variable.

Comment (humblec) [reply pkg/controller/controller_test.go]: @msau42 Thanks! Looks like that.. after untangling the test case , its working as expected!

File: pkg/controller/controller.go (hunk 7 above)
Comment (humblec) [reply pkg/controller/controller.go:522]: Added the test cases ! ..

**Comment by humblec** [comment]
@msau42 I believe this is good to go in!

File: pkg/controller/controller_test.go [hunk 9]
```diff
@@ -1709,9 +1851,88 @@ func runProvisionTest(t *testing.T, k string, tc provisioningTestcase, requested
...
+
+	if tc.expectState == "" {
+		tc.expectState = controller.ProvisioningFinished
+	}
+	if tc.expectState != state {
+		t.Errorf("test %q: expected ProvisioningState %s, got %s", k, tc.expectState, state)
+	}
+
+	if tc.expectedPVSpec != nil {
+		if pv.Name != tc.expectedPVSpec.Name {
```
Comment (msau42) [inline pkg/controller/controller_test.go:1888]: I think the test logic would be cleaner if we did a deepequal of the entire pv spec, instead of matching individual fields one by one. But we can consider cleaning that up separately.

Comment (humblec) [reply pkg/controller/controller_test.go:1888]: Sure, I will do that in a follow-up PR. There are some more places it need to be done in this file anyway.
